File: ocaml/day4/main.ml
open Core

let hd_and_suc xs =
  match xs with x :: y :: _ -> (x, y) | _ -> failwith "Not enough elements"

let as_pair s =
  let as_assignment s' =
    hd_and_suc (String.split s' ~on:'-' |> List.map ~f:Int.of_string)
  in
  hd_and_suc @@ List.map ~f:as_assignment @@ String.split s ~on:','

let is_subsumed ((s, e), (s', e')) = (s >= s' && e <= e') || (s <= s' && e >= e')
let is_overlapping ((s, e), (s', e')) = Int.max s s' <= Int.min e e'

let assignments =
  In_channel.read_lines "../inputs/day4.txt" |> List.map ~f:as_pair

let () =
  let count ~f = Fn.compose List.length (List.filter ~f) in
  printf "Task 1: %d\n" @@ count assignments ~f:is_subsumed;
  printf "Task 1: %d\n" @@ count assignments ~f:is_overlapping

File: rust/day4/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Result as IOResult};

const INPUT_PATH: &str = "../inputs/day4.txt";

type Section = u32;
type Assignment = (Section, Section, Section, Section);

fn as_assignment(line: &str) -> Assignment {
    let sections = line
        .split(&[',', '-'])
        .map(|s| s.parse::<Section>().unwrap())
        .collect::<Vec<_>>();
    (sections[0], sections[1], sections[2], sections[3])
}

fn read_assignments() -> IOResult<Vec<Assignment>> {
    Ok(BufReader::new(File::open(INPUT_PATH)?)
        .lines()
        .map(|s| as_assignment(&s.unwrap()))
        .collect())
}

fn is_subsumed((start1, end1, start2, end2): &&Assignment) -> bool {
    (start1 >= start2 && end1 <= end2) || (start1 <= start2 && end1 >= end2)
}

fn is_overlapping((start1, end1, start2, end2): &&Assignment) -> bool {
    Section::max(*start1, *start2) <= Section::min(*end1, *end2)
}

fn main() -> IOResult<()> {
    let assigns = read_assignments()?;
    println!("Task 1: {}", assigns.iter().filter(is_subsumed).count());
    println!("Task 2: {}", assigns.iter().filter(is_overlapping).count());
    Ok(())
}
